proposals/infra/INF-0005-register-types-and-diagnostics.md

```diff
@@ -130,21 +130,275 @@ is no corresponding member, a warning will be emitted that is treated as an erro
-annotation to determine what diagnostics, if any, to emit.
-`DiagnoseHLSLRegisterAttribute` will be fully responsible for halting compilation
+annotation to determine what diagnostics, if any, to emit. 
+`DiagnoseHLSLRegisterAttribute` will be fully responsible for halting compilation 
 if there is any semantic fault in the application of the register annotation.
 
-
 ## Detailed design
+Firstly, below is some sample code that would cause the proposed diagnostics to be emitted:
+```
+// UDTs with register bindings for resources that don't exist:
+//
+// - warn_hlsl_user_defined_type_missing_member
+// warning: binding type '%select{t|u|b|s|c}0' only applies to types containing '%select{srv resources|uav resources|constant buffer resources|sampler state|constant buffer data}0'
+struct Foo { float f; };
+Foo x : register(t0);
+// warning: binding type 't' only applies to types containing 'srv' resources.
+//
+
+// Mismatched register bindings:
+//
+// - err_hlsl_binding_type_mismatch
+// error: binding type '%select{t|u|b|s|c}0' only applies to '%select{srv resources|uav resources|constant buffer resources|sampler state|constant buffer data}0'
+float f : register(t0);
+// error: binding type 't' only applies to 'srv' resources.
+RWBuffer<float> f : register(c3);
+// error: binding type 'c' only applies to constant buffer data.
+//
+
+// Invalid binding types:
+//
+// err_hlsl_binding_type_invalid
+// error: binding type '%0' is invalid.
+float f : register(x0);
+// error: binding type 'x' is invalid.
+//
+
+// Multiple bindings with the same register type:
+//
+// - err_hlsl_duplicate_register_annotation
+// error: binding type '%select{t|u|b|s|c|i}' cannot be applied more than once
+struct Bar{
+  RWBuffer<int> a;
+  RWBuffer<int> b;
+};
+Bar x : register(t9) : register(t10);
+// error: binding type 'u' cannot be applied more than once
+//
+
+// Binding 'c' when it should be packoffset in non-global scope:
+//
+// - warn_hlsl_register_type_c_packoffset
+// warning: binding type 'c' ignored in buffer declaration. Did you mean 'pack_offset'?
+cbuffer g_cbuffer { float f : register(c2); }
+// warning: binding type 'c' ignored in buffer declaration. Did you mean 'pack_offset'?
+//
+
+// Binding 'b' ignored / being used in the legacy way:
+//
+// - warn_hlsl_deprecated_register_type_b
+// warning: binding type 'b' only applies to constant buffers. The "bool constant" binding type has been removed.
+float f : register(b0);
+// warning: binding type 'b' only applies to constant buffers. The "bool constant" binding type has been removed.
+//
+
+// Binding 'i' ignored (also wording):
+//
+// warn_hlsl_deprecated_register_type_i
+// warning: binding type 'i' ignored. The "integer constant" binding type has been removed
+float f : register(i0);
+// warning: binding type 'i' ignored. The "integer constant" binding type has been removed
+//
+```
+
+In DXC, the analysis and diagnostic emission steps would happen in DiagnoseRegisterType(),

```

Comment:
I think there's too much detail in this section. The list of diagnostics is useful, but I'd rather just review the implementation after we've agreed on the specific diagnostics rather than review both this textual description of it and the implementation itself.

I've not reviewed anything from here until the Behavioral Differences section.